java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/ClientCredentialsGrantTokenRequestPayloadBuilder.java

```diff
@@ -69,48 +80,81 @@
     @SneakyThrows
     public HttpContent buildPayload() {
 
+        //implementation of:
+        // https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#get-a-token
+
+        String oauthClientId = config.getConfigPropertyOrError(OAuthRefreshTokenSourceAuthStrategy.ConfigProperty.CLIENT_ID);
         String tokenEndpoint =
             config.getConfigPropertyOrError(OAuthRefreshTokenSourceAuthStrategy.ConfigProperty.REFRESH_ENDPOINT);
 
-        String oauthClientId = config.getConfigPropertyOrError(OAuthRefreshTokenSourceAuthStrategy.ConfigProperty.CLIENT_ID);
+        Map<String, String> data = new TreeMap<>();
+        //https://datatracker.ietf.org/doc/html/rfc6749#section-4.4.2
+        data.put(PARAM_GRANT_TYPE, getGrantType());
+        config.getConfigPropertyAsOptional(ConfigProperty.TOKEN_SCOPE)
+            .ifPresent(r -> data.put(PARAM_SCOPE, r)); // 'scope' param is optional, per RFC
+
+        //https://datatracker.ietf.org/doc/html/rfc7521#section-4.2
+        data.put(PARAM_CLIENT_ASSERTION_TYPE, CLIENT_ASSERTION_TYPE_JWT);
+        data.put(PARAM_CLIENT_ASSERTION, buildJwtAssertion(oauthClientId, tokenEndpoint));
+        data.put(PARAM_CLIENT_ID, oauthClientId);
+
+        return new UrlEncodedContent(data);
+    }
+
+    /**
+     * build JWT assertion to authenticate client based on config
+     * @param clientId both subject + issuer of token
+     * @param audience for the token (the endpoint/service being called)
+     * @return JWT assertion as a string
+     */
+    @SneakyThrows
+    String buildJwtAssertion(String clientId, String audience) {
 
+        //https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-certificate-credentials
         JsonWebSignature.Header header = new JsonWebSignature.Header();
+        header.setType("JWT");
         header.setAlgorithm("RS256");
-        header.setX509Thumbprint(config.getConfigPropertyOrError(ConfigProperty.PRIVATE_KEY_ID));
-        //header.setType("JWT"); //this IS a JWT, but MSFT examples don't include this
+        header.setX509Thumbprint(encodeKeyId(config.getConfigPropertyOrError(ConfigProperty.PRIVATE_KEY_ID)));
```

Comment:
our main codebase generated certs in powershell, and just took thumbprint value from there ... in proxy, we're using the HEX value as the `PRIVATE_KEY_ID` (which is also the "thumbprint" value that Azure shows in the console for the certificate) - and then per MSFT docs, we have to base64-urlencode that to put it in this JWT header (unclear to me why, given the this value is set on the JWT header, which will be serialized to JSON and then base64-encoded as the first part of the JWT before it's included in the URL ...)